Sqlite rebased

duijf opened the pull request:
This is https://github.com/arianvp/haskell-fido2/pull/15 rebased on the latest `master`.

I tried to preserve the original git history as much as possible, while also making the following changes:

 - Use `Fido2.PublicKey` instead of `Fido2.Ec2Key`
 - Move all HTTP logic into the individual handler functions that we have.
 - Add https://github.com/arianvp/haskell-fido2/commit/4d0597a2c0a64967d2c3f2a5d4096a9e313e1f04 to make the SQLite changes work nicely with the assertion module that I authored in https://github.com/arianvp/haskell-fido2/pull/14
 - Add minor tweaks to `.gitignore` and the Cabal file
 - Clean up a few warnings that resulted from rebasing.
 - Use `Data.Binary.{decode,encode}` before storing the public key parameters in the DB. (Otherwise, these can overflow).
 - Abort database transactions when exceptions occur. (Otherwise the connection remains in the transaction state causing subsequent requests to fail)

This now uses SQLite for the user state. Everything still seems to work now.

ruuda committed:
Store users in database, not in TVar

ruuda committed:
Add getters for listing things from the database

ruuda committed:
Autoformat the database changes

ruuda committed:
Insert user during register-begin already

This way we still have access to the display name and username.

ruuda committed:
Update outdated comment

duijf committed:
cabal: Add Database to other-modules

I was running into a few compile errors which didn't show up because the
Main module wasn't touched. The cause was that cabal didn't want to
recompile the Database module every time I touched it. That is also what
caused some of Ruud's commits to not build.

duijf committed:
git: Ignore SQLite databases

duijf committed:
database: Clarify function name

duijf committed:
server: Fetch credentials from DB during login

duijf committed:
build: Cleanup all warnings

duijf committed:
fido: Add getters for public key fields

ruuda committed:
Add module for interacting with a SQLite database

This is just a toy example database where we can store users and
associated credentials, as an example for the server, don't use
this in production.

duijf committed:
database: Add withTransaction bracket function

duijf committed:
server: User withTransaction over manual begin/commit

This ensures that we always finish our transactions whenever any errors
occur (like UNIQUE constraint violations). Without this, the server
could keep transactions opened meaning subsequent requests would not
work correctly.

duijf committed:
database: Encode Integers before storing in the DB

arianvp commented:
Perhaps we can store the EC2 Points in compressed form (`x` coordinate plus sign) instead of `x/y` coordinates. This eliminates a whole bunch of cryptographic security vulnerabilities by construction

arianvp commented:
lets not make that a blocker. I'll open a new issue for it

arianvp commented on fido/Crypto/Fido2/Protocol.hs:
```diff
@@ -380,6 +385,35 @@ instance Aeson.ToJSON URLEncodedBase64 where
+publicKeyX key =
+  case publicKeyPoint key of
+    ECDSA.Point x _y -> x
+    -- TODO: Does this ever happen?
```
From the COSE docs:

> The encoding of the infinity point is not
        supported.


so no

duijf replied:
We now have a total injection from our own `Point` to `ECDSA.Point`

arianvp commented on server/Database.hs:
```diff
@@ -0,0 +1,187 @@
+      \ values                                                      \
+      \ (?, ?, ?, ?);                                               "
+      ( credentialId,
+        userId,
```
How about we define our own `PublicKey` datatype domain object that is an `x/y` pair and only convert to `ECDSA.Point` on demand when checking the signature. This way we don't need these partial `publicKeyX` functions.

Also it decouples storage from the crypto-implementation; which might help with #22

arianvp replied:
The injection `OurPoint -> ECDSA.Point` will be total, as you never have the `Infinity` case.  

And then you can call `verify` or `sign` or whatever using the constructed point

duijf replied:
I've implemented this now.

arianvp commented on server/Database.hs:
```diff
@@ -0,0 +1,187 @@
+      \ (?, ?, ?, ?);                                               "
+      ( credentialId,
+        userId,
+        Binary.encode $ Fido2.publicKeyX publicKey,
```
Instead of using `Binary` maybe lets use  https://hackage.haskell.org/package/serialise    The PublicKeys we get from Webauthn are in CBOR/COSE format; so it makes sense to store them in CBOR as well?  We can then have a roundtrip test too; which is nice

arianvp replied:
We basically have the `deserialise` function (Though it's a hardcoded mess atm) and just need to add a dual `serialise` function, and then we can write an instance for the `Serialise` typeclass

duijf replied:
I would prefer we make this a non-blocker for merging this, as these encoders / decoders are probably going to change when we use the compressed representation.

Until we get that settled, it doesn't really matter how we store these in the DB right?

ruuda commented:
> Use Data.Binary.{decode,encode} before storing the public key parameters in the DB. (Otherwise, these can overflow).

Overflow in what sense? They were `ByteString` and they were stored as `BLOB`, and the coordinates have a fixed size, so what is there to overflow?

ruuda commented on fido/Crypto/Fido2/Protocol.hs:
```diff
@@ -380,6 +385,35 @@ instance Aeson.ToJSON URLEncodedBase64 where
+    ECDSA.PointO -> error "Got PointO at infinity. We always expect a key with X and Y"
+
+-- Make a public key at the constant curve that we currently only support.
+mkPublicKey :: Integer -> Integer -> PublicKey
```
Shouldn’t this return `Maybe PublicKey`? What if the coordinates don’t lie on the curve?

duijf replied:
I've now renamed this to `mkEcdsaPublicKey` and made it check whether the coordinates are on the curve.

ruuda commented on server/Database.hs:
```diff
@@ -0,0 +1,187 @@
+  pure $ Transaction conn
+
+commit :: Transaction -> IO ()
+commit (Transaction conn) = Sqlite.execute conn "commit;" ()
```
Are these still needed now that there is `withTransaction`?

duijf replied:
Thanks! Removed these

ruuda commented on server/Main.hs:
```diff
@@ -234,7 +220,9 @@ completeLogin sessions users = do
     verifyLogin sessionId session userId challenge = do
       credential <- Scotty.jsonData @(Fido2.PublicKeyCredential Fido2.AuthenticatorAssertionResponse)
-      credentials <- (liftIO $ getUserCredentials userId users) >>= handleError
+      credentials <- liftIO $ Database.withTransaction db $ \tx -> do
```
We already have the credential here, so we can look up only that one by id, right?

duijf replied:
That would require us to trust that the user actually respected the list of `allowedCredentials` that we passed to them. I'm not sure whether that is safe (but I suspect it isn't).

This code ensures that we only check the challenge against the list of credentials that are actually associated with the current user in the DB.

We can also change this to a lookup by both the user ID and the credential ID of course, but I'm not sure if there is a reason for choosing one over the other.

Do you have a preference?

ruuda replied:
> We can also change this to a lookup by both the user ID and the credential ID of course

Yes let's do this.

duijf committed:
protocol: Use our own type for public keys

This allows us to avoid partiality in the publicKeyX and publicKeyY
accessors, as well as expose a smart constructor which always checks
whether the point is valid.

duijf commented:
> Overflow in what sense? They were ByteString and they were stored as BLOB, and the coordinates have a fixed size, so what is there to overflow?

I observed that the signature verification during login didn't work anymore after I got things to compile. When I printed the public keys after reading them back from the DB, they seemed to have overflowed.

But you're right: these fields were `ByteString`s in your branch (I wrongly assumed they were `Integers` there too).

Seems like the `ToField` instance for `Integer` always truncates to SQLite's integer type even though the column is a `BLOB` ([source](https://hackage.haskell.org/package/sqlite-simple-0.4.18.0/docs/src/Database.SQLite.Simple.ToField.html#line-79)). This caused the signature verification to fail after rebasing (which is why I added the call to `Binary.encode`).

duijf committed:
database: Remove manual transaction management functions

duijf committed:
database: Remove unused function

arianvp commented:
Lets merge this "as is" for now. I'm planning to do some major refactoring and I dont want this to become stale again
